engine/access/state_stream/backend.go

```diff
@@ -118,37 +122,23 @@ func New(
 	return b, nil
 }
 
-func (b *StateStreamBackend) getExecutionData(ctx context.Context, blockID flow.Identifier) (*execution_data.BlockExecutionDataEntity, error) {
-	if cached, ok := b.execDataCache.ByID(blockID); ok {
-		b.log.Trace().
-			Hex("block_id", logging.ID(blockID)).
-			Msg("execution data cache hit")
-		return cached, nil
-	}
-	b.log.Trace().
-		Hex("block_id", logging.ID(blockID)).
-		Msg("execution data cache miss")
-
-	seal, err := b.seals.FinalizedSealForBlock(blockID)
-	if err != nil {
-		return nil, fmt.Errorf("could not get finalized seal for block: %w", err)
-	}
-
-	result, err := b.results.ByID(seal.ResultID)
-	if err != nil {
-		return nil, fmt.Errorf("could not get execution result (id: %s): %w", seal.ResultID, err)
+// getExecutionData returns the execution data for the given block height.
+// Expected errors during normal operation:
+// - storage.ErrNotFound or execution_data.BlobNotFoundError: execution data for the given block height is not available.
+func (b *StateStreamBackend) getExecutionData(ctx context.Context, height uint64) (*execution_data.BlockExecutionDataEntity, error) {
+	// fail early if no notification has been received for the given block height.
+	// note: it's possible for the data to exist in the data store before the notification is
+	// received. this ensures a consistent view is available to all streams.
+	if height > b.highestHeight.Load() {
```

Comment:
how do we ensure there is no missing gap in heights below the `highestHeight`?

In order to ensure that we need to only increment `highestHeight` if `highestHeight + 1` is received, right?

Reply:
this is handled by the requester engine. The requester sends notifications when execution data for a new highest consecutive height is received

Reply:
I added a comment about this requirement to `OnExecutionData` in [41d2c97](https://github.com/onflow/flow-go/pull/4383/commits/41d2c971bed4da0cf0482111aa17933b454ef0df)